Add support for configuring output file

stepanstipl opened the pull request:
This PR adds support for `-O` flag to configure output to a file to main binary and all the existing printers (json, text).

Related changes introduced:
- Increased test coverage for the common code and existing printers in the `printer` package
- Refactored Main tests to pass arguments via env variable, this was required to maintain consistency between runs (tmp dir)


Fixes #304

_Sonar CLoud complains about test func naming with underscores (which are generated by Goland and questionably a standard for unexported func tests in Go) and some repeated strings all in tests. I plan to ignore it... in the longer term, we should really get to enabling some linters in a way that makes sense), the SC is getting annoying._

dark0dave commented on cmd/kubent/main.go:
```diff
@@ -154,10 +154,11 @@ func main() {
 	}
 
-	printer, err := printer.NewPrinter(config.Output)
+	printer, err := printer.NewPrinter(config.Output, config.OutputFile)
```
shouldnt this be only passed if the flag is set?

stepanstipl replied:
I would say no, given go doesn't really do optional args, we just pass the default (which is stdout `-`), same as we do just a couple of lines above, e.g. `loadedRules, err := rules.FetchRegoRules(additionalKinds)`.

dark0dave commented on cmd/kubent/main.go:
```diff
@@ -154,10 +154,11 @@ func main() {
 	if err != nil {
 		log.Fatal().Err(err).Msg("Failed to create printer")
 	}
+	defer printer.Close()
```
defer should be handled inside the printer

dark0dave replied:
and only if the flag has been set

dark0dave commented on pkg/config/config.go:
```diff
@@ -81,3 +91,20 @@ func validateAdditionalResources(resources []string) error {
+		return fmt.Errorf("output file name can't be empty (use - for stdout)")
+	}
+
+	if outputFileName != "-" {
```
shouldn't this return and error? I would catch this above

if outputFileName == ""  or outputFileName == "-"

dark0dave commented on cmd/kubent/main.go:
```diff
@@ -154,10 +154,11 @@ func main() {
 	if err != nil {
 		log.Fatal().Err(err).Msg("Failed to create printer")
 	}
+	defer printer.Close()
```
this should happen inside of printer

stepanstipl replied:
I don't know how to. Go doesn't have a concept of destructors, and `defer Close()` kind of a standard, see for example `io.Closer` [^1], and it's users responsibility to call it.

[^1]: https://pkg.go.dev/io#Closer

dark0dave commented on pkg/config/config.go:
```diff
@@ -81,3 +91,20 @@ func validateAdditionalResources(resources []string) error {
 	return nil
 }
+
+// validateOutputFile checks if output file name is valid and if the
```
do we need the comments? I feel the name of the function is suffiecent

stepanstipl replied:
Comments are standard for generating docs using `godoc`. And while it typically does not display unexported funcs, it does have a switch to do that. So for that reason, I would prefer to have, hopefully meaningful, comments as a standard.

I agree though that these are not very helpful, so would be happy for better suggestions :).

dark0dave replied:
ah, if they are generated, then we should like generate them for the whole repo. Rather than just these changes.

stepanstipl replied:
I don't mean the comments themselves are generated, but that these are used in generated documentation. E.g. a few times mentions [^1] is generated from [^2]. I agree that at some time I would like to have comments for all funcs, probably when we put in some linter to check for that, but for now, I would prefer to add those rather than not.


[^1]: https://pkg.go.dev/io#Closer
[^2]: https://cs.opensource.google/go/go/+/refs/tags/go1.19.1:src/io/io.go;l=100-106

stepanstipl replied:
That btw. brings up a very good question - we do build on many platforms, incl. Windows. Should we run tests on all of those as well?

 _I would say yes, but not sure how difficult is the implementation._

dark0dave commented on pkg/printer/printer.go:
```diff
@@ -2,31 +2,75 @@ package printer
+	outputFile *os.File
+}
+
+// newCommonPrinter creates new printer that prints to given output file
```
these comments are not very useful

stepanstipl replied:
Better suggestion? Explained above why I would like to have comments.

dark0dave commented on pkg/printer/printer.go:
```diff
@@ -2,31 +2,75 @@ package printer
 }
+
+// ensureOutputFile will open file for writing, or create one if non-existent
+func ensureOutputFile(fileName string) (*os.File, error) {
```
ensureOutputFile ensureOutputFileExists

dark0dave commented on pkg/printer/printer.go:
```diff
@@ -2,31 +2,75 @@ package printer
+		return os.Stdout, nil
+	}
+
+	of, err := os.OpenFile(fileName, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
```
does this work for windows?

dark0dave replied:
might want to use ModeAppend instead

stepanstipl replied:
This is commonly used in examples. I'm not sure though, haven't tested on Windows. Why shouldn't it?

dark0dave replied:
0644 is a unix term

stepanstipl replied:
I assume it should work just fine - https://pkg.go.dev/os#FileMode:
> A FileMode represents a file's mode and permission bits. The bits have the same definition on all systems, so that information about files can be moved from one system to another portably. Not all bits apply to all systems. ...

stepanstipl replied:
That btw. brings up a very good question - we do build on many platforms, incl. Windows. Should we run tests on all of those as well?

I would say yes, but not sure how difficult is the implementation.

dark0dave reviewed (changes requested):
I think my main take away is that we should not be closing a file if we are using standard out, feels incorrect

stepanstipl commented:
Thanks for reviewing.

> I think my main take away is that we should not be closing a file if we are using standard out, feels incorrect.

We're not, this is handled within the relevant printer's `Close()` implementation. In general, given lack of destructors in Go, "Closer" interface is a common pattern in go, e.g. see the previously mentioned `io.Closer` [^1], os `database/sql` package [^2].

But don't think about this just in the context of closing file, but as a generic destruct/cleanup interface to free any resources, and that any printer can implement as they wish. E.g. let's say we implement a "printer" to Cloud Logging, or to a database, this would be a place to teardown any open clients and connections.

[^1]: https://pkg.go.dev/io#Closer
[^2]: https://pkg.go.dev/database/sql@go1.19.1#Conn.Close

dark0dave commented on cmd/kubent/main_test.go:
```diff
@@ -211,3 +236,22 @@ func TestGetServerVersion(t *testing.T) {
 	}
 }
+
+func encodeBase64(args []string) (string, error) {
```
I think these are helpers?

stepanstipl replied:
Yep.

dark0dave commented:
> Thanks for reviewing.
> 
> > I think my main take away is that we should not be closing a file if we are using standard out, feels incorrect.
> 
> We're not, this is handled within the relevant printer's `Close()` implementation. In general, given lack of destructors in Go, "Closer" interface is a common pattern in go, e.g. see the previously mentioned `io.Closer` [1](#user-content-fn-1-dd815946a879eb610c29af2e231f789d), os `database/sql` package [2](#user-content-fn-2-dd815946a879eb610c29af2e231f789d).
> 
> But don't think about this just in the context of closing file, but as a generic destruct/cleanup interface to free any resources, and that any printer can implement as they wish. E.g. let's say we implement a "printer" to Cloud Logging, or to a database, this would be a place to teardown any open clients and connections.
> ## Footnotes
> 
>     1. https://pkg.go.dev/io#Closer [leftwards_arrow_with_hook](#user-content-fnref-1-dd815946a879eb610c29af2e231f789d)
> 
>     2. https://pkg.go.dev/database/sql@go1.19.1#Conn.Close [leftwards_arrow_with_hook](#user-content-fnref-2-dd815946a879eb610c29af2e231f789d)

Sorry I worded that badly I don't mind defer.Close() at all, I am talking about closing the printer in main. Its not very obvious why we'd close the printer in main if we are not doing any file writing. Additionally its a different level of abstraction which probably should probably not be handled in main.

stepanstipl commented:
> Sorry I worded that badly I don't mind defer.Close() at all, I am talking about closing the printer in main. Its not very obvious why we'd close the printer in main if we are not doing any file writing. Additionally its a different level of abstraction which probably should probably not be handled in main.

Well, currently that's where the printer is created [^1]. I.e. typically `defer` follows after object creation, I'm not sure where else to put it (it also has to wait till we're done with the printer).

What would you suggest?

[^1]: https://github.com/doitintl/kube-no-trouble/blob/8c7bc37e1d3ae458327dfc420fef4397a1b016ff/cmd/kubent/main.go#L157

stepanstipl commented:
Renamed `ensureOutputFile` -> `ensureOutputFileExists` and moved printer code in main to its own function `outputResults` with tests.

@dark0dave I think it's ready for another look. Thanks.

dark0dave reviewed (approved):
LGTM

dark0dave commented:
@stepanstipl please rebase and fix conflicts
